### PR TITLE
Greengrass Ipc 1.0

### PR DESCRIPTION
**Pull request by bretambrose**
* Initial release of Greengrass Ipc 1.0

Does not include sample which will be submitted once the release is tagged and available in npm.
User guide may have some broken links until docs are generated by cutting a release.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by sbSteveK — APPROVED** [review]
Looks good. Trivial note: we add "This file is generated" to the header under the copyright/license on generated files. Some of these have the look of generated files and we may want to put that up there if they are.

File: GreengrassIPC.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+This guide is a work in progress.  Feedback is welcomed.
+
+## Developer Preview Disclaimer
+Javascript Greengrass IPC support is currently in **developer preview**.  We encourage feedback at all times, but feedback during the preview window
+is especially valuable in shaping the final product.  During the preview period we may make backwards-incompatible changes to the
+public API.
+
+## Cross-SDK Differences
+The Javascript Greengrass IPC implementation has several differences relative to the IPC implementations in the other v2 SDKs:
```
Comment (TwistedTwigleg) [inline GreengrassIPC.md:34]: Extremely minor nitpick: `in the other v2 SDKs` -> `in the other v2 IoT SDKs`

File: GreengrassIPC.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+* No handler types - Javascript Greengrass IPC uses Javascript's event system (events, listeners, emitters) for all asynchronous notifications.
+* Single client - Other SDKs have both a v1 and v2 client, where the v1 client is low-level and more verbose (and complex) in usage, while the v2 client is streamlined and less prone to accidental misuse.  Javascript Greengrass IPC skips the v1 API and starts directly with the v2 client approach.
```
Comment (TwistedTwigleg) [inline GreengrassIPC.md:36]: Two extremely small nitpicks:

* `Other SDKs have both...` -> `Other v2 IoT SDKs have both...`
* `Javascript Greengrass IPC skips the v1 API and starts...` -> `Javascript Greengrass IPC skips the v1 client approach and starts...`

File: GreengrassIPC.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+be closed.
+
+Once a client has been closed, it cannot be used further.
+
+### Client Events
+The Greengrass IPC client provides a single event that can be listened to, the `disconnection` event.  A `disconnection`
+is emitted when the underlying connection is torn down.  A `disconnection` event will only be emitted if a call to 
+`connect` previously succeeded; connection establishment failures only manifest as a rejected promise.
+
+Calling `close` on a connected client will cause a `disconnection` event to be emitted as part of the shutdown process.
```
Comment (TwistedTwigleg) [inline GreengrassIPC.md:85]: Question: Is there a way to tell the difference between a disconnection event caused by calling `close` versus a disconnection event caused by the client becoming disconnected from Greengrass IPC?

File: GreengrassIPC.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+operations with streaming output.
+
+A streaming operation is not performed automatically when the client API is invoked (otherwise the user would not have
+an opportunity to attach event listeners beforehand, potentially missing streaming events).  Instead the client API
+creates and returns a `StreamingOperation` object.
+
+A `StreamingOperation` object supports several events:
+* `message` - event emitted every time a message is successfully received on the stream from the server.  The data included with this event is specific to each streaming operation type.
+* `streamError` - event emitted when either the deserialization of an incoming message failed, or the deserialization resulted in a modeled service error.  The data included with this event is an [RpcError](https://aws.github.io/aws-iot-device-sdk-js-v2/node//classes/eventstream_rpc.RpcError.html).
+* `ended` - event emitted when the stream has been closed for any reason.  Only emitted for operations that were successfully activated.
```
Comment (TwistedTwigleg) [inline GreengrassIPC.md:127]: Nitpick: add `prior to stream close` after `activated`.

File: GreengrassIPC.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    // handle incoming publishes
+    streamingSubscription.on("message", (message: greengrasscoreipc.model.IoTCoreMessage) => {
+        // handle the received MQTT publish message
+    });
+    
+    // start the stream
+    await streamingSubscription.activate();
+```
+
+Like a Greengrass IPC client, a streaming operation must be closed when finished with.  The Greengrass IPC client will
```
Comment (TwistedTwigleg) [inline GreengrassIPC.md:148]: Nitpick: Add `with the close() function` after `.. finished with`

**Review by TwistedTwigleg — APPROVED** [review]

File: GreengrassIPC.md (hunk 3 above)
Comment (bretambrose) [reply GreengrassIPC.md:85]: The associated CrtError will be "user-initiated disconnect"

File: GreengrassIPC.md (hunk 4 above)
Comment (bretambrose) [reply GreengrassIPC.md:127]: You can't activate a stream after close so it may be a little redundant

**Pull request merged**